File: shopping-cart/python/src/shopping_cart/discount_policy.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import Protocol

from .money import Money
from .quantity import Quantity

MIN_PERCENT = 0
MAX_PERCENT = 100
HUNDRED_PERCENT = Decimal(100)


class DiscountPolicy(Protocol):
    def apply(self, unit_price: Money, quantity: Quantity) -> Money: ...


class NoDiscount:
    def apply(self, unit_price: Money, quantity: Quantity) -> Money:
        return unit_price * quantity.value


NO_DISCOUNT = NoDiscount()


@dataclass(frozen=True)
class PercentOff:
    percent: int

    def __post_init__(self) -> None:
        if not isinstance(self.percent, int) or isinstance(self.percent, bool):
            raise ValueError("Percent must be an integer between 0 and 100")
        if self.percent < MIN_PERCENT or self.percent > MAX_PERCENT:
            raise ValueError("Percent must be an integer between 0 and 100")

    def apply(self, unit_price: Money, quantity: Quantity) -> Money:
        gross = unit_price.amount * quantity.value
        multiplier = (HUNDRED_PERCENT - Decimal(self.percent)) / HUNDRED_PERCENT
        return Money(gross * multiplier)


@dataclass(frozen=True)
class FixedOff:
    amount: Money

    def __post_init__(self) -> None:
        if self.amount < Money.zero():
            raise ValueError("Fixed discount amount must not be negative")

    def apply(self, unit_price: Money, quantity: Quantity) -> Money:
        gross = unit_price * quantity.value
        discounted = gross - self.amount
        return Money.zero() if discounted < Money.zero() else discounted


@dataclass(frozen=True)
class BuyXGetY:
    buy_count: int
    free_count: int

    def __post_init__(self) -> None:
        if not isinstance(self.buy_count, int) or self.buy_count <= 0:
            raise ValueError("Buy count must be a positive integer")
        if not isinstance(self.free_count, int) or self.free_count <= 0:
            raise ValueError("Free count must be a positive integer")

    def apply(self, unit_price: Money, quantity: Quantity) -> Money:
        group_size = self.buy_count + self.free_count
        full_groups, remainder = divmod(quantity.value, group_size)
        chargeable = full_groups * self.buy_count + min(remainder, self.buy_count)
        return unit_price * chargeable


@dataclass(frozen=True)
class BulkPricing:
    threshold: Quantity
    bulk_unit_price: Money

    def __post_init__(self) -> None:
        if self.bulk_unit_price < Money.zero():
            raise ValueError("Bulk unit price must not be negative")

    def apply(self, unit_price: Money, quantity: Quantity) -> Money:
        effective = self.bulk_unit_price if quantity.value >= self.threshold.value else unit_price
        return effective * quantity.value

File: shopping-cart/typescript/src/DiscountPolicy.ts
import { Money } from './Money.js';
import { Quantity } from './Quantity.js';

export interface DiscountPolicy {
  apply(unitPrice: Money, quantity: Quantity): Money;
}

export class NoDiscount implements DiscountPolicy {
  static readonly instance = new NoDiscount();
  apply(unitPrice: Money, quantity: Quantity): Money {
    return unitPrice.times(quantity.value);
  }
}

const MIN_PERCENT = 0;
const MAX_PERCENT = 100;
const HUNDRED_PERCENT = 100;

export class PercentOff implements DiscountPolicy {
  constructor(public readonly percent: number) {
    if (!Number.isInteger(percent) || percent < MIN_PERCENT || percent > MAX_PERCENT) {
      throw new RangeError('Percent must be an integer between 0 and 100');
    }
  }

  apply(unitPrice: Money, quantity: Quantity): Money {
    const multiplier = (HUNDRED_PERCENT - this.percent) / HUNDRED_PERCENT;
    return new Money(unitPrice.amount * quantity.value * multiplier);
  }
}

export class FixedOff implements DiscountPolicy {
  constructor(public readonly amount: Money) {
    if (amount.lessThan(Money.zero)) {
      throw new RangeError('Fixed discount amount must not be negative');
    }
  }

  apply(unitPrice: Money, quantity: Quantity): Money {
    const gross = unitPrice.times(quantity.value);
    const discounted = gross.minus(this.amount);
    return discounted.lessThan(Money.zero) ? Money.zero : discounted;
  }
}

export class BuyXGetY implements DiscountPolicy {
  constructor(public readonly buyCount: number, public readonly freeCount: number) {
    if (!Number.isInteger(buyCount) || buyCount <= 0) {
      throw new RangeError('Buy count must be a positive integer');
    }
    if (!Number.isInteger(freeCount) || freeCount <= 0) {
      throw new RangeError('Free count must be a positive integer');
    }
  }

  apply(unitPrice: Money, quantity: Quantity): Money {
    const groupSize = this.buyCount + this.freeCount;
    const fullGroups = Math.floor(quantity.value / groupSize);
    const remainder = quantity.value % groupSize;
    const chargeable = fullGroups * this.buyCount + Math.min(remainder, this.buyCount);
    return unitPrice.times(chargeable);
  }
}

export class BulkPricing implements DiscountPolicy {
  constructor(public readonly threshold: Quantity, public readonly bulkUnitPrice: Money) {
    if (bulkUnitPrice.lessThan(Money.zero)) {
      throw new RangeError('Bulk unit price must not be negative');
    }
  }

  apply(unitPrice: Money, quantity: Quantity): Money {
    const effectiveUnitPrice = quantity.value >= this.threshold.value ? this.bulkUnitPrice : unitPrice;
    return effectiveUnitPrice.times(quantity.value);
  }
}
